README
http://hackage.haskell.org/package/rex

Provides a quasi-quoter for regular expressions which
yields a tuple, of appropriate arity and types,
representing the results of the captures.  Allows the user 
to specify parsers for captures as inline Haskell.  Can 
also be used to provide typeful pattern matching in
function definitions and case patterns.

To build / install:

./Setup.hs configure --user
./Setup.hs build
./Setup.hs install

See the haddock or Text/Regex/PCRE/QQT.hs for documentation.


CHANGES
0.1: Jul 25 2011
  * initial release

0.2: Sep 24 2011
  * Added custom configuration of PCRE options.
  * Added non-precompiling quasiquoters.
  * Fixed a bug where patterns with no captures would fail.
  * Decided to remove the defaulting to 'read' - too much magic.

0.3: Sep 25 2011
  * Fixed a capture indexing bug, where capture fields which aren't bound would
    cause subsequent bound captures to be incorrect.
  * Above bug fix actually neatened up code.
  * Added configuration of default mapping pattern.


Test.hs
{-# LANGUAGE TemplateHaskell, QuasiQuotes, ViewPatterns #-}

module Test where

import Text.Regex.PCRE.Rex

import qualified Data.ByteString.Char8 as B
import Data.Maybe (catMaybes, isJust)

math x = mathl x 0

mathl [] x = x
mathl [rex|^  \s*(?{ read -> y }\d+)\s*(?{ s }.*)$|] x = mathl s y
mathl [rex|^\+\s*(?{ read -> y }\d+)\s*(?{ s }.*)$|] x = mathl s $ x + y
mathl [rex|^ -\s*(?{ read -> y }\d+)\s*(?{ s }.*)$|] x = mathl s $ x - y
mathl [rex|^\*\s*(?{ read -> y }\d+)\s*(?{ s }.*)$|] x = mathl s $ x * y
mathl [rex|^ /\s*(?{ read -> y }\d+)\s*(?{ s }.*)$|] x = mathl s $ x / y
mathl str x = error str

peano :: String -> Maybe Int
peano = [rex|^(?{ length . filter (=='S') } \s* (?:S\s+)*Z)\s*$|]

parsePair :: String -> Maybe (String, String)
parsePair = [rex|^<\s* (?{ }[^\s,>]+) \s*,\s* (?{ }[^\s,>]+) \s*>$|]

-- From http://www.regular-expressions.info/dates.html
parseDate :: String -> Maybe (Int, Int, Int)
parseDate [rex|^(?{ read -> y }(?:19|20)\d\d)[- /.]
                (?{ read -> m }0[1-9]|1[012])[- /.]
                (?{ read -> d }0[1-9]|[12][0-9]|3[01])$|]
  |  (d > 30 && (m `elem` [4, 6, 9, 11]))
  || (m == 2 &&
       (d ==29 && not (mod y 4 == 0 && (mod y 100 /= 0 || mod y 400 == 0)))
    || (d > 29)) = Nothing
  | otherwise = Just (y, m, d)
parseDate _ = Nothing

onNull a f [] = a
onNull _ f xs = f xs

nonNull = onNull Nothing

disjunct [rex| ^(?:(?{nonNull $ Just . head -> a} .)
             | (?{nonNull $ Just . head -> b} ..)
             | (?{nonNull $ Just . last -> c} ...))$|] =
  head $ catMaybes [a, b, c]

btest = [brex|(?{}hello)|] (B.pack "hello") == Just (B.pack "hello")


test.hs
{-# LANGUAGE TemplateHaskell, QuasiQuotes, ViewPatterns #-}

import Text.Regex.PCRE.QQT
import Debug.Trace

-- mathr ([$reg|({ -> Just x}\d+)\s*\+\s*({id -> y}.+)|]) = x + mathr y

debug x = trace (show x) x

math ([reg|^(\d+)\s*({id}.*)|]       -> (Just y, s)) x = math (debug s) y
math ([reg|^\+\s*(\d+)\s*({id}.*)$|] -> (Just y, s)) x = math (debug s) $ x + y
math ([reg|^-\s*(\d+)\s*({id}.*)$|]  -> (Just y, s)) x = math (debug s) $ x - y
math ([reg|^\*\s*(\d+)\s*({id}.*)$|] -> (Just y, s)) x = math (debug s) $ x * y
math ([reg|^/\s*(\d+)\s*({id}.*)$|]  -> (Just y, s)) x = math (debug s) $ x / y
math [] x = x

{-
*Main> math "2 + 2" 0
4.0
*Main> math "2 + 2 * 5" 0
20.0
*Main> math "2 + 20 * 5" 0
110.0
*Main> math "2 + 20 * 5 / 3" 0
36.666666666666664
-}


Text/Regex/PCRE/QQT.hs
{-# LANGUAGE TemplateHaskell, QuasiQuotes, TupleSections, ParallelListComp, ViewPatterns #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Text.Regex.PCRE.QQT
-- Copyright   :  Michael Sloan 2011
--
-- Maintainer  :  Michael Sloan (mgsloan@gmail.com)
-- Portability :  unportable
-- Stability   :  experimental
-- 
-- This module provides a template Haskell quasiquoter for regular
-- expressions, which provides the following features:
-- 
--   1) Compile-time checking that the regular expression is valid.
-- 
--   2) Arity of resulting tuple based on the number of capture patterns in
--   the regular expression.
--
--   3) By default utilizes type inference to determine how to parse capture
--   patterns - uses "maybeRead" function, which yields a "(Read a) => Just a"
--   value on successful parse.
--
--   4) Allows for the inline interpolation of a mapping function String -> a.
--
-- Inspired by / copy-modified from Matt Morrow's regexqq package:
-- http://hackage.haskell.org/packages/archive/regexqq/latest/doc/html/src/Text-Regex-PCRE-QQ.html
-- And code from Erik Charlebois's interpolatedstring-qq package:
-- http://hackage.haskell.org/packages/archive/interpolatedstring-qq/latest/doc/html/Text-InterpolatedString-QQ.html

module Text.Regex.PCRE.QQT (reg, makeExpr, maybeRead, pack, unpack) where

import Text.Regex.PCRE.Light (Regex,PCREOption,PCREExecOption)
import qualified Text.Regex.PCRE.Light as PCRE

import Control.Monad (liftM, join)

import qualified Data.ByteString as B
import Data.ByteString.Internal (c2w,w2c)
import Data.Either.Utils (forceEitherMsg)
import Data.List (groupBy, inits, sortBy)
import Data.List.Split (split, onSublist)
import Data.Maybe (catMaybes, listToMaybe)
import Data.Ord (comparing)
import Data.Char (isSpace)

import Debug.Trace

import Language.Haskell.TH
import Language.Haskell.TH.Quote
import Language.Haskell.Meta.Parse

type ParseChunk = Either String (Int, String)
type ParseChunks = (Int, String, [(Int, String)])

{- TODO: idea - provide a variant which allows splicing in an expression
  that evaluates to regex string.  The tricky thing here is that this
  splice might contain capture groups - thus, the parser here would need to
  be referenced in order to dispatch 
-}

{- TODO: target Text.Regex.Base -}

-- | The regular expression quasiquoter.
reg :: QuasiQuoter
reg = QuasiQuoter
        (checkRegex makeExpr . parseIt)
        (checkRegex makePat . parseIt)
        undefined undefined

-- Gives an error at compile time if the regex string is invalid.
checkRegex :: (ParseChunks -> a) -> ParseChunks -> a
checkRegex f x@(_, pat, _) = 
  forceEitherMsg ("Error compiling regular expression [$reg|"  ++ pat ++ "|]")
    (PCRE.compileM (pack pat) pcreOpts) `seq` f x


-- Template Haskell Code Generation
-------------------------------------------------------------------------------

-- Creates the template haskell Exp which corresponds to the parsed interpolated
-- regex.  This particular code mainly just handles making "maybeRead" the
-- default for captures which lack a parser definition, and defaulting to making
-- the parser that doesn't exist
makeExpr :: ParseChunks -> ExpQ
makeExpr (cnt, pat, exs) = buildExpr cnt pat
    . map (processExpr . snd . head)
    . groupSortBy (comparing fst)
    $ exs ++ ((0, "") : [(i, "maybeRead") | i <- [1..cnt]])
  
-- Creates the template haskell Pat which corresponds to the parsed interpolated
-- regex. As well as handling the aforementioned defaulting considerations, this
-- turns per-capture view patterns into a single tuple-resulting view pattern.
-- 
-- E.g. [reg| ... ({e1 -> v1} ...) ... ({e2 -> v2} ...) ... |] becomes
--      [reg| ... ({e1} ...) ... ({e2} ...) ... |] -> (v1, v2)
makePat :: ParseChunks -> PatQ
makePat (cnt, pat, exs) = do
  viewExp <- buildExpr cnt pat $ map (liftM fst) ys
  return . ViewP viewExp . TupP . map snd $ catMaybes ys
 where
  ys = map (join . liftM floatBoth . processView . snd . head)
     . groupSortBy (comparing fst)
     $ (0, "") : exs ++ [(i, "") | i <- [1..cnt]]

  processView "" = Nothing
  processView xs = case splitFromBack 2 ((split . onSublist) "->" xs) of
    (_, [r]) -> onSpace Nothing (Just . (processExpr "maybeRead",) . processPat) r
    (concat -> l, [_, r]) -> Just (processExpr (onSpace "maybeRead" id $ l), processPat r)
    (_, _) -> Nothing

-- Here's where the main meat of the template haskell is generated.  Given the
-- number of captures, the pattern string, and a list of capture expressions,
-- yields the template Haskell Exp which parses a string into a tuple.
buildExpr :: Int -> String -> [Maybe Exp] -> ExpQ
buildExpr cnt pat hexps = do
  vx <- newName "x"
  emptyE <- [|""|]
  caseExp <-  [| fmap (map unpack)
               $ PCRE.match (regex pat) (pack $(return $ VarE vx)) pcreExecOpts |]
  let mkMatch rcnt xs = Match
        (if null xs then ConP (mkName "Nothing") []
                    else ConP (mkName "Just") . single . ListP $ map VarP xs)
        (NormalB . TupE . map (uncurry AppE) . catMaybes
                 . zipWith (curry floatFst) hexps 
                 $ (map VarE xs) ++ replicate rcnt emptyE)
        []
  return . LamE [VarP vx] . CaseE caseExp . zipWith mkMatch [cnt+1,cnt..]
         $ inits [mkName $ "r" ++ show i | i <- [0..cnt]]

-- Parse a Haskell expression into a template Haskell Exp, yielding Nothing for
-- strings which just consist of whitespace.
processExpr :: [Char] -> Maybe Exp
processExpr xs = onSpace Nothing
  (Just . forceEitherMsg ("Error while parsing capture mapper " ++ xs) .  parseExp) xs

-- Parse a Haskell pattern match into a template Haskell Pat, yielding Nothing for
-- strings which just consist of whitespace.
processPat :: [Char] -> Maybe Pat
processPat xs = onSpace Nothing
  (Just . forceEitherMsg ("Error while parsing capture mapper " ++ xs) .  parsePat) xs

-- Parsing
-------------------------------------------------------------------------------

-- Postprocesses the results of the chunk-wise parse output, into the pattern to
-- be pased to the regex engine, and the interpolated 
parseIt :: String -> ParseChunks
parseIt xs = (cnt, concat [x | Left x <- results]
             , [(i, x) | Right (i, x) <- results])
  where (cnt, results) = parseRegex (filter (`notElem` "\r\n") xs) "" True 0

-- A pair of mutually-recursive functions, one for processing the quotation
-- and the other for the anti-quotation.

parseRegex :: String -> String -> Bool -> Int -> (Int, [ParseChunk])
parseRegex inp s b ix = case inp of
  -- Disallow branch-reset capture.
  ('(':'?':'|':_) ->
    error "Branch reset pattern (?| not allowed in fancy quasi-quoted regex."

 
  -- Ignore non-capturing parens / handle backslash escaping
  ('\\':'\\'  :xs) -> parseRegex xs ("\\\\" ++ s) False ix
  ('\\':'('   :xs) -> parseRegex xs (")\\"  ++ s) False ix
  ('\\':')'   :xs) -> parseRegex xs ("(\\"  ++ s) False ix
  ('(':'?':':':xs) -> parseRegex xs (":?("  ++ s) False ix

  -- Anti-quote for processing a capture group.
  ('(':'{':xs) -> mapSnd ((Left $ reverse ('(':s)) :)
                    $ parseHaskell xs "" (ix+1)
                    
  -- Anti-quote for processing the whole match. Only applies at the beginning.
  ('{':xs) -> if b then parseHaskell xs "" ix
                   else parseRegex xs ('{':s) False ix

  -- Keep track of how many capture groups we've seen.
  ('(':xs) -> parseRegex xs ('(':s) False (ix+1)

  -- Consume the regular expression contents.
  (x:xs) -> parseRegex xs (x:s) False ix
  [] -> (ix, [Left $ reverse s])

parseHaskell :: String -> String -> Int -> (Int, [ParseChunk])
parseHaskell inp s ix = case inp of
  -- Escape } in the Haskell splice using a backslash.
  ('\\':'}':xs) -> parseHaskell xs ('}':s) ix

  -- Capture accumulated antiquote, and continue parsing regex literal.
  ('}':xs) -> mapSnd ((Right (ix, reverse s)):)
            $ parseRegex xs "" False ix

  -- Consume the antiquoute contents, appending to a reverse accumulator.
  (x:xs) -> parseHaskell xs (x:s) ix
  [] -> error "Regular-expression Haskell splice is never terminated with a trailing }"

-- TODO: provide bytestring variant.

-- Utils
-------------------------------------------------------------------------------

-- The following 2 functions are referenced in the generated TH code, as well as
-- this module.

pack :: String -> B.ByteString
pack = B.pack . fmap c2w

unpack :: B.ByteString -> String
unpack = fmap w2c . B.unpack

-- Compiles a regular expression with the default options.
regex :: String -> Regex
regex = flip PCRE.compile pcreOpts . pack

-- | The default read definition - yields "Just x" when there is a valid parse,
--   and Nothing otherwise.
maybeRead :: (Read a) => String -> Maybe a
maybeRead = fmap fst . listToMaybe . reads


-- TODO: allow for a bundle of parameters to be passed in at the beginning
-- of the quasiquote. Would also be a juncture for informing arity /
-- extension information.

pcreOpts :: [PCREOption]
pcreOpts =
  [ PCRE.extended
  , PCRE.multiline ]
  -- , dotall, caseless, utf8
  -- , newline_any, PCRE.newline_crlf ]

pcreExecOpts :: [PCREExecOption]
pcreExecOpts = []
  -- [ PCRE.exec_newline_crlf
  -- , exec_newline_any, PCRE.exec_notempty
  -- , PCRE.exec_notbol, PCRE.exec_noteol ]

single :: a -> [a]
single x = [x]

groupSortBy :: (a -> a -> Ordering) -> [a] -> [[a]]
groupSortBy f = groupBy (\x -> (==EQ) . f x) . sortBy f

debug :: (Show a) => a -> a
debug x = trace (show x) x

mapFst :: (a -> c) -> (a, b) -> (c, b)
mapFst f (x, y) = (f x, y)

mapSnd :: (b -> c) -> (a, b) -> (a, c)
mapSnd f (x, y) = (x, f y)

floatFst :: (Functor f) => (f a, b) -> f (a, b)
floatFst (x, y) = fmap (,y) x

floatSnd :: (Functor f) => (a, f b) -> f (a, b)
floatSnd (x, y) = fmap (x,) y

floatBoth :: (Monad m) => (m a, m b) -> m (a, b)
floatBoth (x, y) = do
  x' <- x
  y' <- y
  return (x', y')

splitFromBack :: Int -> [a] -> ([a], [a])
splitFromBack i xs = (reverse b, reverse a)
  where (a, b) = splitAt i $ reverse xs

dropAllBut :: Int -> [a] -> [a]
dropAllBut i = reverse . take i . reverse

onSpace :: a -> (String -> a) -> String -> a
onSpace x f s | all isSpace s = x
              | otherwise = f s

mapLeft :: (a -> a') -> Either a b -> Either a' b
mapLeft f = either (Left . f) Right

mapRight :: (b -> b') -> Either a b -> Either a b'
mapRight f = either Left (Right . f)

{-
--TODO: use something like this to cache compiled regex.
-- Even better would be to do the compilation step compile time :)
-- http://stackoverflow.com/questions/141650/how-do-you-make-a-generic-memoize-function-in-haskell
memoize :: Ord a => (a -> b) -> (a -> b)
memoize f = unsafePerformIO $ do 
    r <- newIORef Map.empty
    return $ \ x -> unsafePerformIO $ do 
        m <- readIORef r
        case Map.lookup x m of
            Just y  -> return y
            Nothing -> do 
                    let y = f x
                    writeIORef r (Map.insert x y m)
                    return y
-}


Text/Regex/PCRE/Precompile.hs
{-# LANGUAGE TupleSections #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Text.Regex.PCRE.Rex
-- Copyright   :  (c) Michael Sloan 2011
--
-- Maintainer  :  Michael Sloan (mgsloan@gmail.com)
-- Stability   :  unstable
-- Portability :  unportable
-- 
-- This module provides support for accessing the compiled representation of
-- PCRE regular expressions.  This byte array encodes a lookup-table based
-- representation for the regular expression, and can be safely written
-- out and read back by compatible PCRE versions.
--
-----------------------------------------------------------------------------

module Text.Regex.PCRE.Precompile where

import Control.Monad (liftM)

import qualified Data.ByteString.Char8 as B
import Data.ByteString.Unsafe (unsafePackCStringLen, unsafeUseAsCString)

import Foreign.ForeignPtr (withForeignPtr, newForeignPtr_)
import Foreign.Ptr (nullPtr, castPtr)
import Foreign.Marshal (alloca)
import Foreign.Storable (peek)
import Foreign.C.Types (CInt)

import Text.Regex.PCRE.Light
import Text.Regex.PCRE.Light.Base

-- | A type synonym indicating which ByteStrings represent PCRE-format compiled
-- data.
type CompiledBytes = B.ByteString

-- | Compiles the given regular expression, and assuming nothing bad
-- happens, yields the bytestring filled with PCRE's compiled
-- representation.
precompile :: B.ByteString -> [PCREOption] -> IO (Maybe CompiledBytes)
precompile pat opts = regexToTable $ compile pat opts

-- | Takes a compiled regular expression, and yields .
regexToTable :: Regex -> IO (Maybe CompiledBytes)
regexToTable (Regex p _) =
  withForeignPtr p $ \pcre -> alloca $ \res -> do
    success <- c_pcre_fullinfo pcre nullPtr info_size res
    len <- return . fromIntegral =<< (peek res :: IO CInt)
    if success >= 0 
      then withForeignPtr p (liftM Just . unsafePackCStringLen . (, len) . castPtr)
      else return Nothing

-- | Creates a regular expression from the compiled representation.
regexFromTable :: CompiledBytes  -> IO Regex
regexFromTable bytes = unsafeUseAsCString bytes $ \cstr -> do
   ptr <- newForeignPtr_ (castPtr cstr)
   return $ Regex ptr bytes


Text/Regex/PCRE/Rex.hs
{-# LANGUAGE TemplateHaskell, QuasiQuotes, TupleSections, ViewPatterns #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Text.Regex.PCRE.Rex
-- Copyright   :  (c) Michael Sloan 2011
--
-- Maintainer  :  Michael Sloan (mgsloan@gmail.com)
-- Stability   :  unstable
-- Portability :  unportable
-- 
-- This module provides a template Haskell quasiquoter for regular
-- expressions, which provides the following features:
-- 
-- 1) Compile-time checking that the regular expression is valid.
-- 
-- 2) Arity of resulting tuple based on the number of selected capture patterns
-- in the regular expression.
--
-- 3) Allows for the inline interpolation of mapping functions :: String -> a.
--
-- 4) Precompiles the regular expression at compile time, by calling into the
-- PCRE library and storing a 'B.ByteString' literal representation of its state.
--
-- 5) Compile-time configurable to use different PCRE options, turn off
-- precompilation, use 'B.ByteString's, or set a default mapping expression.
--
-- Since this is a quasiquoter library that generates code using view patterns,
-- the following extensions are required:
--
-- > {-# LANGUAGE TemplateHaskell, QuasiQuotes, ViewPatterns #-}
--
-- Here's a silly example which parses peano numbers of the form Z, S Z,
-- S S Z, etc.  The \s+ means that it is not sensitive to the quantity or type
-- of seperating whitespace. (these examples can also be found in Test.hs)
--
-- > peano :: String -> Maybe Int
-- > peano = [rex|^(?{ length . filter (=='S') } \s* (?:S\s+)*Z)\s*$|]
--
-- > *Main> peano "Z"
-- > Just 0
-- > *Main> peano "S Z"
-- > Just 1
-- > *Main> peano "S   S Z"
-- > Just 2
-- > *Main> peano "S S S Z"
-- > Just 3
-- > *Main> peano "invalid"
-- > Nothing
--
-- The token \"(?{\" introduces a capture group which has a mapping applied to
-- the -- result - in this case \"length . filter (=='S')\".  If the ?{ ... }
-- are omitted, then the capture group is not taken as part of the results of
-- the match.  If the contents of the ?{ ... } is omitted, then 'id' is assumed:
--
-- > parsePair :: String -> Maybe (String, String)
-- > parsePair = [rex|^<\s* (?{ }[^\s,>]+) \s*,\s* (?{ }[^\s,>]+) \s*>$|]
--
-- The following example is derived from
-- http://www.regular-expressions.info/dates.html
--
-- > parseDate :: String -> Maybe (Int, Int, Int)
-- > parseDate [rex|^(?{ read -> y }(?:19|20)\d\d)[- /.]
-- >                 (?{ read -> m }0[1-9]|1[012])[- /.]
-- >                 (?{ read -> d }0[1-9]|[12][0-9]|3[01])$|]
-- >   |  (d > 30 && (m `elem` [4, 6, 9, 11]))
-- >   || (m == 2 &&
-- >       (d == 29 && not (mod y 4 == 0 && (mod y 100 /= 0 || mod y 400 == 0)))
-- >      || (d > 29)) = Nothing
-- >   | otherwise = Just (y, m, d)
-- > parseDate _ = Nothing
--
-- The above example makes use of the regex quasi-quoter as a pattern matcher.
-- The interpolated Haskell patterns are used to construct an implicit view
-- pattern out of the inlined ones.  The above pattern is expanded to the
-- equivalent:
--
-- > parseDate ([rex|^(?{ read }(?:19|20)\d\d)[- /.]
-- >                  (?{ read }0[1-9]|1[012])[- /.]
-- >                  (?{ read }0[1-9]|[12][0-9]|3[01])$|]
-- >           -> Just (y, m, d))
--
--
-- Caveat: Since haskell-src-exts does not support parsing view-patterns, the
-- above is implemented as a relatively naive split on \"->\".  It presumes that
-- the last \"->\" in the interpolated pattern seperates the pattern from an
-- expression on the left.  This allows for lambdas to be present in the
-- expression, but prevents nesting view patterns.
--
-- There are also a few other inelegances:
--
-- 1) PCRE captures, unlike .NET regular expressions, yield the last capture
-- made by a particular pattern.  So, for example, (...)*, will only yield one
-- match for '...'.  Ideally these would be detected and yield an implicit [a].
--
-- 2) Patterns with disjunction between captures ((?{f}a) | (?{g}b)) will
-- provide the empty string to one of f / g.  In the case of pattern
-- expressions, it would be convenient to be able to map multiple captures into
-- a single variable / pattern, preferring the first non-empty option.  The
-- general logic for this is a bit complicated, and postponed for a later
-- release.
-- 
-- 3) While this error is believed to no longer exist, the following error
-- 
-- >  <interactive>: out of memory (requested 17584491593728 bytes)
--
-- Used to occur when evaluating in GHCi, due to a bug in the way precompilation
-- worked.  If this happens, please report it, and as a temporary work around,
-- make your own quasiquoter using \"'rexConf' _ False _ _ _\" to disable
-- pre-compilation.
--
-- Since pcre-light is a wrapper over a C API, the most efficient interface is
-- ByteStrings, as it does not natively speak Haskell lists.  The [rex| ... ]
-- quasiquoter implicitely packs the input into a bystestring, and unpacks the
-- results to strings before providing them to your mappers.  The 'brex'
-- 'QuasiQuoter' is provided for this purpose.  You can also define your own
-- 'QuasiQuoter' - the definitions of the default configurations are as follows:
--
-- > rex  = rexConf False True "id" rexPCREOptions []
-- > brex = rexConf True  True "id" rexPCREOptions []
--
-- As mentioned, the other Bool determines whether precompilation is used.  The
-- string following is the default mapping expression, used when omitted.
-- Due to GHC staging restrictions, your configuration will need to be in a
-- different module than its usage.
--
-- Inspired by Matt Morrow's regexqq package:
-- <http://hackage.haskell.org/packages/archive/regexqq/latest/doc/html/src/Text-Regex-PCRE-QQ.html>
--
-- And code from Erik Charlebois's interpolatedstring-qq package:
-- <http://hackage.haskell.org/packages/archive/interpolatedstring-qq/latest/doc/html/Text-InterpolatedString-QQ.html>
--
-----------------------------------------------------------------------------

module Text.Regex.PCRE.Rex (
  rex, brex, rexConf, rexPCREOptions,
  maybeRead, padRight, makeQuasiMultiline) where

import qualified Text.Regex.PCRE.Light as PCRE
import qualified Text.Regex.PCRE.Light.Base as PCRE
import Text.Regex.PCRE.Precompile

import Control.Arrow (first)
import Control.Monad (liftM)

import qualified Data.ByteString.Char8 as B
import Data.Either.Utils (forceEitherMsg)
import Data.List (find)
import Data.List.Split (split, onSublist)
import Data.Maybe (catMaybes, listToMaybe, fromJust, isJust)
import Data.Maybe.Utils (forceMaybeMsg)
import Data.Char (isSpace)

import Language.Haskell.TH
import Language.Haskell.TH.Quote
import Language.Haskell.Meta.Parse

import System.IO.Unsafe (unsafePerformIO)

{- TODO:
  * Fix mentioned caveats
  * Target Text.Regex.Base ? 
  * Provide a variant which allows splicing in an expression that evaluates to
    regex string.
  * Add unit tests
  * Consider allowing passing arity lists in to configuration
-}

type ParseChunk = Either String (Int, String)
type ParseChunks = (Int, String, [(Int, String)])
type Config = (Bool, Bool, String, [PCRE.PCREOption], [PCRE.PCREExecOption])

-- | Default regular expression quasiquoter for 'String's and 'B.ByteString's,
-- respectively.
rex, brex :: QuasiQuoter
rex  = rexConf False True "id" rexPCREOptions []
brex = rexConf True  True "id" rexPCREOptions []

-- | This is a 'QuasiQuoter' transformer, which allows for a whitespace-sensitive
-- quasi-quoter to be broken over multiple lines.  The default 'rex' and
-- 'brex' functions do not need this as they are already whitespace insensitive.
-- However, if you create your own configuration, which omits the 'PCRE.extended'
-- parameter, then this could be useful. The leading space of each line is
-- ignored, and all newlines removed.
makeQuasiMultiline :: QuasiQuoter -> QuasiQuoter
makeQuasiMultiline (QuasiQuoter a b c d) =
  QuasiQuoter (a . pre) (b . pre) (c . pre) (d . pre)
 where
  pre = concat . (\(x:xs) -> x : map (dropWhile isSpace) xs) . lines

-- | Default compilation time PCRE options.  The default is 'PCRE.extended',
-- which causes whitespace to be nonsemantic, and ignores # comments.
rexPCREOptions :: [PCRE.PCREOption]
rexPCREOptions = [ PCRE.extended ]
  -- , dotall, caseless, utf8
  -- , newline_any, PCRE.newline_crlf ]

-- | A configureable regular-expression QuasiQuoter.  Takes the options to pass
-- to the PCRE engine, along with 'Bool's to flag 'B.ByteString' usage and
-- non-compilation respecively.  The provided 'String' indicates which mapping
-- function to use, when one is omitted - \"(?{} ...)\".
rexConf :: Bool -> Bool -> String -> [PCRE.PCREOption] -> [PCRE.PCREExecOption]
        -> QuasiQuoter
rexConf bs pc d os eos = QuasiQuoter
        (makeExp conf . parseIt)
        (makePat conf . parseIt)
        undefined undefined
 where
  conf = (bs, pc, d, [PCRE.combineOptions os], [PCRE.combineExecOptions eos])

-- Template Haskell Code Generation
-------------------------------------------------------------------------------

-- Creates the template haskell Exp which corresponds to the parsed interpolated
-- regex.  This particular code mainly just handles making "read" the
-- default for captures which lack a parser definition, and defaulting to making
-- the parser that doesn't exist
makeExp :: Config -> ParseChunks -> ExpQ
makeExp conf (cnt, pat, exs) = buildExp conf cnt pat exs'
 where
  exs' = map (\ix -> liftM (processExp conf . snd) $ find ((==ix).fst) exs) [0..cnt]
  
-- Creates the template haskell Pat which corresponds to the parsed interpolated
-- regex. As well as handling the aforementioned defaulting considerations, this
-- turns per-capture view patterns into a single tuple-resulting view pattern.
-- 
-- E.g. [reg| ... (?{e1 -> v1} ...) ... (?{e2 -> v2} ...) ... |] becomes
--      [reg| ... (?{e1} ...) ... (?{e2} ...) ... |] -> (v1, v2)
makePat :: Config -> ParseChunks -> PatQ
makePat conf (cnt, pat, exs) = do
  viewExp <- buildExp conf cnt pat $ map (liftM fst) views
  return . ViewP viewExp
         . (\xs -> ConP (mkName "Just") [TupP xs])
         . map snd $ catMaybes views
 where
  views :: [Maybe (Exp, Pat)]
  views = map (\ix -> liftM (processView . snd) $ find ((==ix).fst) exs) [0..cnt]

  processView :: String -> (Exp, Pat)
  processView xs = case splitFromBack 2 ((split . onSublist) "->" xs) of
    (_, [r]) -> onSpace r (error $ "blank pattern in view: " ++ r)
                          ((processExp conf "",) . processPat)
    -- View pattern
    (l, [_, r]) -> (processExp conf $ concat l, processPat r)
    -- Included so that Haskell doesn't warn about non-exhaustive patterns
    -- (even though the above are exhaustive in this context)
    _ -> undefined

-- Here's where the main meat of the template haskell is generated.  Given the
-- number of captures, the pattern string, and a list of capture expressions,
-- yields the template Haskell Exp which parses a string into a tuple.
buildExp :: Config -> Int -> String -> [Maybe Exp] -> ExpQ
buildExp (bs, nc, _, pcreOpts, pcreExecOpts) cnt pat xs =
  [| let r = $(get_regex) in
     $(process) . (flip $ PCRE.match r) $(liftRS pcreExecOpts)
   . $(if bs then [| id |] else [| B.pack |]) |]
 where
  liftRS x = [| read shown |]
   where shown = show x

  --TODO: make sure this takes advantage of bytestring fusion stuff - is
  -- the right pack / unpack. Or use XOverloadedStrings
  get_regex = if nc
    then [| unsafePerformIO (regexFromTable $! $(table_bytes)) |]
    else [| PCRE.compile (B.pack pat) $(liftRS pcreOpts)|]
  table_bytes = [| B.pack $(runIO table_string) |]
  table_string = precompile (B.pack pat) pcreOpts >>= 
    return . LitE . StringL . B.unpack . 
    forceMaybeMsg "Error while getting PCRE compiled representation\n"

  process = case (null vs, bs) of
    (True, _)  -> [| liftM ( const () ) |]
    (_, False) -> [| liftM (($(return maps)) . padRight "" pad . map B.unpack) |]
    (_, True)  -> [| liftM (($(return maps)) . padRight B.empty pad) |]
  pad = cnt + 2
  maps = LamE [ListP . (WildP:) $ map VarP vs]
       . TupE . map (uncurry AppE)
       -- filter out all "Nothing" exprs
       . map (first fromJust) . filter (isJust . fst)
       -- [(Expr, Variable applied to)]
       . zip xs $ map VarE vs
  vs = [mkName $ "v" ++ show i | i <- [0..cnt]]

-- Parse a Haskell expression into a template Haskell Exp
processExp :: Config -> String -> Exp
processExp (_, _, d, _, _) xs
  = forceEitherMsg ("Error while parsing capture mapper `" ++ xs ++ "'")
  . parseExp $ onSpace xs d id

-- Parse a Haskell pattern match into a template Haskell Pat, yielding Nothing
-- for patterns which consist of just whitespace.
processPat :: String -> Pat
processPat xs
  = forceEitherMsg ("Error while parsing capture pattern `" ++ xs ++ "'")
  $ parsePat xs

-- Parsing
-------------------------------------------------------------------------------

-- Postprocesses the results of the chunk-wise parse output, into the pattern to
-- be pased to the regex engine, and the interpolated 
parseIt :: String -> ParseChunks
parseIt xs = ( cnt, concat [x | Left x <- results]
             , [(i, x) | Right (i, x) <- results])
 where
  (cnt, results) = parseRegex (filter (`notElem` "\r\n") xs) "" (-1)

-- A pair of mutually-recursive functions, one for processing the quotation
-- and the other for the anti-quotation.

-- TODO: add check for erroneous { }

parseRegex :: String -> String -> Int -> (Int, [ParseChunk])
parseRegex inp s ix = case inp of
  -- Disallow branch-reset capture.
  ('(':'?':'|':_) ->
    error "Branch reset pattern (?| not allowed in quasi-quoted regex."

  -- Ignore non-capturing parens / handle backslash escaping.
  ('\\':'\\'  :xs) -> parseRegex xs ("\\\\" ++ s) ix
  ('\\':'('   :xs) -> parseRegex xs (")\\"  ++ s) ix
  ('\\':')'   :xs) -> parseRegex xs ("(\\"  ++ s) ix
  ('(':'?':':':xs) -> parseRegex xs (":?("  ++ s) ix

  -- Anti-quote for processing a capture group.
  ('(':'?':'{':xs) -> mapSnd ((Left $ reverse ('(':s)) :)
                    $ parseHaskell xs "" (ix + 1)

  -- Keep track of how many capture groups we've seen.
  ('(':xs) -> parseRegex xs ('(':s) (ix + 1)

  -- Consume the regular expression contents.
  (x:xs) -> parseRegex xs (x:s) ix
  [] -> (ix, [Left $ reverse s])

parseHaskell :: String -> String -> Int -> (Int, [ParseChunk])
parseHaskell inp s ix = case inp of
  -- Escape } in the Haskell splice using a backslash.
  ('\\':'}':xs) -> parseHaskell xs ('}':s) ix

  -- Capture accumulated antiquote, and continue parsing regex literal.
  ('}':xs) -> mapSnd ((Right (ix, reverse s)):)
            $ parseRegex xs "" ix

  -- Consume the antiquoute contents, appending to a reverse accumulator.
  (x:xs) -> parseHaskell xs (x:s) ix
  [] -> error "Rex haskell splice terminator, }, never found"

-- Utils
-------------------------------------------------------------------------------

-- | A possibly useful utility function - yields 'Just' x when there is a
-- valid parse, and 'Nothing' otherwise.
maybeRead :: (Read a) => String -> Maybe a
maybeRead = fmap fst . listToMaybe . reads

splitFromBack :: Int -> [a] -> ([a], [a])
splitFromBack i xs = (reverse b, reverse a)
  where (a, b) = splitAt i $ reverse xs

onSpace :: String -> a -> (String -> a) -> a
onSpace s x f | all isSpace s = x
              | otherwise = f s

-- | Given a desired list-length, if the passed list is too short, it is padded
-- with the given element.  Otherwise, it trims.
padRight :: a -> Int -> [a] -> [a]
padRight _ 0 xs = xs
padRight v i [] = replicate i v
padRight v i (x:xs) = x : padRight v (i-1) xs

mapSnd :: (t -> t2) -> (t1, t) -> (t1, t2)
mapSnd f (x, y) = (x, f y)
